connectors/src/connectors/microsoft/temporal/activities.ts

```diff
@@ -26,21 +46,258 @@ export async function fullSyncActivity({
-      dataSourceConfig
-    )}\nExample message: ${JSON.stringify(message)}`
+export async function syncOneFile(
+  client: Client,
+  {
+    connectorId,
+    dataSourceConfig,
+    file,
+    parent,
+    startSyncTs,
+    isBatchSync = false,
+  }: {
+    connectorId: ModelId;
+    dataSourceConfig: DataSourceConfig;
+    file: microsoftgraph.DriveItem;
+    parent: MicrosoftRootResource;
+    startSyncTs: number;
+    isBatchSync?: boolean;
+  }
+) {
+  const localLogger = logger.child({
+    provider: "microsoft",
+    connectorId: parent.connectorId,
+    internalId: file.id,
+    name: file.name,
+  });
+
+  if (!file.file) {
+    throw new Error(`Item is not a file: ${JSON.stringify(file)}`);
+  }
+
+  const itemApiPath = getDriveItemApiPath(
+    file,
+    microsoftInternalIdFromNodeData(parent)
   );
+
+  const documentId = microsoftInternalIdFromNodeData({
+    itemApiPath,
+    nodeType: "file",
+  });
+
+  const fileResource =
+    await MicrosoftNodeResource.fetchByInternalId(documentId);
+
+  // Early return if lastSeenTs is greater than workflow start.
+  // This allows avoiding resyncing already-synced documents in case of activity failure
+  if (
+    fileResource?.lastSeenTs &&
+    fileResource.lastSeenTs > new Date(startSyncTs)
+  ) {
+    return true;
+  }
+
+  if (fileResource?.skipReason) {
+    localLogger.info(
+      { skipReason: fileResource.skipReason },
+      "Skipping file sync"
+    );
+    return false;
+  }
+
+  const url =
+    "@microsoft.graph.downloadUrl" in file
+      ? file["@microsoft.graph.downloadUrl"]
+      : null;
+
+  if (!url) {
+    localLogger.error("Unexpected missing download URL");
+    throw new Error("Unexpected missing download URL");
+  }
+
+  if (!url) {
+    localLogger.info("No download URL found");
+    return false;
+  }
+
+  // If the file is too big to be downloaded, we skip it.
+  if (file.size && file.size > MAX_FILE_SIZE_TO_DOWNLOAD) {
+    localLogger.info("File size exceeded, skipping file.");
+
+    return false;
+  }
+
+  const downloadRes = await axios.get(`${url}`, {
+    responseType: "arraybuffer",
+  });
+
+  if (downloadRes.status !== 200) {
+    localLogger.error(
+      `Error while downloading file ${file.name}: ${downloadRes.status}`
+    );
+    throw new Error(
+      `Error while downloading file ${file.name}: ${downloadRes.status}`
+    );
+  }
+
+  async function getDocumentContent() {
+    try {
+      const converted = await mammoth.convertToHtml({
+        buffer: Buffer.from(downloadRes.data),
+      });
+
+      const extracted = new turndown()
+        .remove(["style", "script", "iframe", "noscript", "form", "img"])
+        .turndown(converted.value);
+
+      return extracted.trim();
+    } catch (err) {
+      localLogger.error(
+        {
+          error: err,
+        },
+        `Error while converting docx document to text`
+      );
+      throw err;
+    }
+  }
+
+  const documentContent = await getDocumentContent();
```

Comment:
in theory, if the document is a docx (or any ms file) we don't need to use mammoth - the api directly extract html content - you should be able to use `drives/{{drive-id}}/items/{{file-id}}/content?format=html` (it probably won't work on all files, only those created with 365)

Reply:
As discussed IRL, I had seen on [this doc](https://learn.microsoft.com/en-us/graph/api/driveitem-get-content-format?view=graph-rest-1.0&tabs=http#format-options) that this doesn't work for .docx (only work to convert to pdf)
I also had tried just in case, but it didn't work